scipy/stats/stats.py

```diff
@@ -5718,3 +5720,170 @@ def rankdata(a, method='average'):
 
     # average method
     return .5 * (count[dense] + count[dense - 1] + 1)
+
+
+def ratio_unif(f, size=1, x1=None, x2=None, y2=None, shift=0, maxiter=None,
+               bounds=None):
+    """
+    Generate random variables using ratio-of-uniforms method.
+
+    Given a continuous univariate density f and a constant c, define the set
+    A(c) := {(x, y) : 0 < y <= sqrt(f(x/y + c)) }. If (U, V) is a random
+    vector uniformly distributed over A(c), then U/V + c follows a distribution
+    with density f.
+
+    The above result (see [1]_, [2]_) can be used to sample random variables
+    using only the density. Typical choices of c are zero or the mode of f.
+    The set A(c) is a subset of the rectangle R = [-x1, x2] x [0, y2] where
+    - x1 = inf (x - c) sqrt(f(x))
+    - x2 = sup (x - c) sqrt(f(x))
+    - y2 = sup sqrt(f(x))
+
+    In particular, these values are finite if f is bounded and x**2 f(x) is
+    bounded (i.e. subquadratic tails). One can generate (U, V) uniformly on
+    R and return U/V + c if (U, V) are also in A(c) which can be directly
+    verified. Intuitively, this works well if A(c) fills up most of the
+    enclosing rectangle such that the probability is high that (U, V) lies in
+    A(c) whenever it lies in R(c) as the number of required iterations
+    becomes too large otherwise.
+
+    Parameters
+    ----------
+    f : callable
+        Scalar function defining the density.
+    size : int, optional
+        The number of random variables to generate.
+    x1 : float, optional
+        The lower bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    x2 : float, optional
+        The upper bound of the bounding rectangle in the x-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    y2 : float, optional
+        The upper bound of the bounding rectangle in the y-direction.
+        Default is None. In that case, attempt to find bound numerically.
+    maxiter : int, optional
+        Maximum number of iterations to generate the random variables.
+        Default is None. In that case, maxiter is set to a value that
+        ensures successful simulation with a high probability.
+    bounds : tuple of floats (a, b), optional
+        Default is none. In case any of the boundaries of the bounding
+        rectangle is not specified, attempt to find missing boundary
+        numerically over the range specified by bounds.
+
+    Returns
+    -------
+    rvs : array
+        array with the random variables distributed according to density f
+
+    Notes
+    -----
+    If the bounding rectangle is not correctly specified (i.e. if it does not
+    contain A(c)), the algorithm samples from a distribution different from
+    the one given by the density f. It is therefore recommended to perform a
+    test such as `stats.kstest` as a check.
```

Comment:
Any comment on the time / space complexity of this algorithm -- especially given that there are academic citations below?

Reply:
I added a comment on expected number of iterations.